### ISSUE TITLE
Deleting a word copies that word to the clipboard (or crashes trying)

### ISSUE THREAD

**Issue by jaraco** [xonsh/xonsh#4227]
## xonfig

<details>

```
$ xonfig
+------------------+----------------------+
| xonsh            | 0.9.27               |
| Git SHA          | 71fe9014             |
| Commit Date      | Jan 29 08:58:58 2021 |
| Python           | 3.9.2                |
| PLY              | 3.11                 |
| have readline    | True                 |
| prompt toolkit   | 3.0.18               |
| shell type       | prompt_toolkit       |
| pygments         | None                 |
| on posix         | True                 |
| on linux         | False                |
| on darwin        | True                 |
| on windows       | False                |
| on cygwin        | False                |
| on msys2         | False                |
| is superuser     | False                |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
| on jupyter       | False                |
| jupyter kernel   | None                 |
| xontrib 1        | vox                  |
| xontrib 2        | voxapi               |
+------------------+----------------------+
```

</details>

## Expected Behavior
Deleting a word shouldn't mutate the clipboard.

## Current Behavior
Hitting Option+Delete deletes the last word, but also copies it to the clipboard. When a pyperclip clipboard isn't available, causes a traceback. See prompt-toolkit/python-prompt-toolkit#1412 for more details on the crash.

## Steps to Reproduce
Install latest `xonsh[ptk]` on macOS. Type a word `foo` into the shell. Hit Option+Delete (I'm unsure the equivalent key binding for other platforms). The word is now gone. Now hit Cmd+V and `foo` appears on the terminal. Whatever was on the clipboard before is now lost.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by jaraco** [comment]
Looks like #3989 may have introduced this behavior.

**Comment by anki-code** [comment]
@jnoortheen please take a look.

**Comment by jnoortheen** [comment]
@jaraco could you post the error when pyperclip is not installed?

**Comment by jaraco** [comment]
> @jaraco please post the error when pyperclip is not installed.

Sorry I wasn't clear. Pyperclip is installed - there's just no viable clipboard.

I see now the issue may be that the reason pyperclip is installed is because I used `pipx inject xonsh jaraco.clipboard`. I do that because my .xonshrc imports `jaraco.clipboard`.

I think a suitable workaround to suppress the crash is to uninstall `jaraco.clipboard` in environments that don't have a clipboard. Ideally, however, the clipboard support would be robust to having `pyperclip` installed but non-viable.

That still leaves the issue of environments where the clipboard is viable but the mutating of the clipboard is undesirable.



>     2\. disable copy on deleting word and other such kill-ring feature implemented by ptk via a variable.

I contend this behavior should be opt-in and not opt-out. Though I guess it's arguable that the presence of `pyperclip` is something of an opt-in. In my case, I'm installing pyperclip not to opt in to this functionality, but for other purposes.

**Comment by jnoortheen** [comment]
We can make the Clipboard class passed to ptk as configurable to solve this. One can also implement their own interface and pass in this way. @jaraco are you interested in opening a PR?

**Comment by jaraco** [comment]
Yes, perhaps. Do you have any more information about this "kill-ring feature"?

**Comment by jnoortheen** [comment]
They are activated by the delete sequence shortcuts here https://github.com/prompt-toolkit/python-prompt-toolkit/blob/master/prompt_toolkit/key_binding/bindings/emacs.py

**Comment by daniel-shimon** [comment]
Hi!
The 'copy-on-delete' thing is REALLY annoying and I agree it should be opt-in.
The code responsible is actually in https://github.com/prompt-toolkit/python-prompt-toolkit/blob/master/prompt_toolkit/key_binding/bindings/named_commands.py in the functions `kill_line`, `kill_word`, `unix_word_rubout` and `unix_line_discard`.
All of them run `event.app.clipboard.set_text(deleted)`. So as I see it we have two options:

1. Create a PR in `prompt_toolkit` to make this optional somehow
2. Meanwhile, we can implement a Clipboard that actively checks if we're in any of these functions, and if so ignores the `set_text` invocation. This is super hacky but is probably a good solution until we introduce the change inside prompt_toolkit

**Comment by daniel-shimon** [comment]
Hey @jaraco do you intend to make a PR for this issue?
If not I can happily do so :)

**Comment by jaraco** [comment]
Hey Daniel. I'd probably eventually get to it, but I'm more than happy for someone else to work on it. Go for it! Let me know if I can help.
